tools/ir-generator/irclass.cpp

```diff
@@ -662,6 +664,42 @@ void IrField::generate_impl(std::ostream &) const {
 
 ////////////////////////////////////////////////////////////////////////////////////
 
+void IrVariantField::resolve() {
+    for (const Type *type : *types) resolveType(type);
+}
+
+void IrVariantField::generate(std::ostream &out, bool asField) const {
+    if (asField) {
+        out << IrClass::indent << "using " << name << "_variant = std::variant<";
+        bool first = true;
+        for (const Type *type : *types) {
+            if (!first) out << ", ";
+
+            // FIXME: Support variant of IR node pointers
+            // const IrClass *cls = type->resolve(clss ? clss->containedIn : nullptr);
+            // if (cls != nullptr) out << "const ";
```

Comment:
Yes. The problem is "inline" feature of fields that changes the types. So, in the sense variants are always "inline". However, if someone would want to have default short-cut logic at some point, this is the placeholder to implement. Likely this would require some syntax extension (e.g. allowing "inline" on variant types)

Reply:
The `inline` feature of fields is there to disable the automatic conversion of them to (const) pointers that normally happens in the ir-generator.

Reply:
The question is: consider one would want `variant` of several node pointers. Then one would need to spell out `variant<const IR::Expression *, const IR::Argument *>` explicitly. as `variant<Expression, Argument>` will be translated into `variant<IR::Expression, IR::Argument>` storing the nodes inline. Is this desired? Or we can have e.g. `variant<inline Expression, inline Argument>` to be translated into `variant<IR::Expression, IR::Argument>` and `variant<Expression, Argument>` into `variant<const IR::Expression *, const IR::Argument *>` correspondingly?

Reply:
You can now explicitly write `variant<const IR::Expression *, const IR::Argument *>` and that should work.

We could add support for `inline` to variants (currently, you'll just get a syntax error), but then where you currently have `variant<Expression, Argument>` would need to change to `variant<inline Expression, inline Argument>` to have the behavior you currently have.

Reply:
Right. So my question is do we want to support this? Or leave as-is until the first real usecase?

Reply:
I think for consistency we should support it, but it is ok to defer it for awhile.

Reply:
Ok, then what would be the preferred syntax? `inline` on the inner type?

Reply:
I think this applies not only to variant, but also to things like `optional` and in general to any templated type (except for `IR::Vector` and `IR::IndexedVector` which take T, but the value type is in fact `const T *`. I believe `optional` now behaves just as `variant`, i.e. `std::optional<int> fld` is translated verbatim.

BTW. is the transformation from `T` to `const T *` implicit only for IR types? Because one could expect that `std::optional<int> fld` would be translated as `const std::optional<int> *fld` but it is not. On the other hand, having to write `inline std::optional<inline int>` does not sound very nice either.

Reply:
@vlstill `optional` does not generate `std::optional`. It just generates a version of constructor that does not initialize this field. And the field is default-initialized.

The transformation is only for IR types, yes and as far as I can see, only in the "outer" position, e.g. for field types but not for template parameters, etc.

Reply:
I meant explicitly using `std::optional<T>` in the .def file, not the `optional` specifier :-). Just as another example similar to `variant`. `std::optional` can be useful for inline fields that hold non-IR types and are optional so I've noticed the transformation does not occur there. I can't think of a reason to use an IR type in `std::optional`, so if the "pointer-adding" only works for IR types then this example is not that important.

Reply:
Oh, yes. Note that for `Vector<Node>` the transformation is implicit as the storage is always `std::vector<const Node*>`. So, maybe we'll just keep the things as-is, and return when the real usecase will appear?

Reply:
Yes. I think that it makes sense to state it like this:

> Given a field definition `Type fldName` if the type of the field (`Type`) is an IR node type and the `inline` specifier is not used, then the real generated field is `const Type *fldName`, otherwise it is `Type fldName`.

This should be the current behavior and I think it is rather consistent and easy to use. If anyone has a use case to use IR types in nested position *and* wants to use pointers there, they will need to use the pointer explicitly. I think that is not too much cost. Also, if we tried to make the transformation implicit for IR types appearing anywhere, we would need to make a exception for `IR::Vector`  and `IR::IndexedVector`.

Reply:
But it seems the IR types are actually transformed in nested positions in some cases :-/. I.e. `Util::Enumerator<IDeclaration> *getDeclarations() const override` actually means `Util::Enumerator<const IR::IDeclaration *> *IR::Function::getDeclarations() const`. Anyway, this is beyond this PR, I am fine with types nested inside `variant` not being lifted now.